src/tests/integration_tests/test_usage_tracker.py

```diff
@@ -0,0 +1,160 @@
+    db.check_usage(token_amount="2000", user_id="jack")
+    check_usage = db.check_usage(token_amount="10", user_id="jack")
+    assert check_usage["can_execute"] == False
+
+    # check if the usage tracker has been updated after 4 seconds
+    timer = threading.Timer(4.0, (lambda: ""))
+    timer.start()
+    timer.join()
+    check_usage = db.check_usage(token_amount="10", user_id="jack")
+    assert check_usage["can_execute"] == True
+
+
+@pytest.fixture
+def mock_download_from_s3(monkeypatch):
+    # by copying and renaming this function mocks downloading from sqlight file
+    def mock_download(*args, **kwargs):
+        # sqlight db file path which acts as s3 bucket
+        dormant_test_db_path = "./tests/data/test_usage_counter.db"
+
+        # testing db file path
+        active_test_db_path = "./test_usage.db"
+
+        # Copy and rename the file
+        shutil.copyfile(dormant_test_db_path, active_test_db_path)
+
+    monkeypatch.setattr(UserUsageTracker, "download_from_s3", mock_download)
+    
+@pytest.fixture
+def mock_upload_to_s3_from_s3(monkeypatch):
+    # by copying and renaming the testing sqlight db file to the location of the dormant file this function mocks uploading to s3 bucket.
+    def mock_upload(*args, **kwargs):
+        # sqlight db file path which acts as s3 bucket
+        dormant_test_db_path = './tests/data/test_usage_counter.db'
+        check = os.path.exists(dormant_test_db_path)
+        if check:
+            os.remove(dormant_test_db_path)
+
+        # testing db file path
+        active_test_db_path = "./test_usage.db"
+
+        # Copy and rename the file
+        shutil.copyfile(active_test_db_path, dormant_test_db_path)
+
+    monkeypatch.setattr(UserUsageTracker, "upload_to_s3", mock_upload)
+
+
+def test_download_db_from_s3(mock_download_from_s3):
+    check = os.path.exists("test_usage.db")
+    if check:
+        os.remove("test_usage.db")
+    UserUsageTracker.download_from_s3(
+        db_name="test_usage.db", db_url="sqlite:///test_usage.db"
+    )
+    third_check = os.path.exists("test_usage.db")
+    assert third_check, "downloading sqlight file has failed"
+
+
+def test_whitelist_from_s3(mock_download_from_s3 , mock_upload_to_s3_from_s3):
+    check = os.path.exists("test_usage.db")
+    if not check:
+        UserUsageTracker.download_from_s3(
+            db_name="test_usage.db", db_url="sqlite:///test_usage.db"
+        )
+    db = UserUsageTracker(db_name="test_usage.db", db_url="sqlite:///test_usage.db")
+    db.add_to_whitelist(user_id=user_id)
+    db.upload_to_s3()
+    os.remove("test_usage.db")
+    UserUsageTracker.download_from_s3(
+        db_name="test_usage.db", db_url="sqlite:///test_usage.db"
+    )
+    db2 = UserUsageTracker(db_name="test_usage.db", db_url="sqlite:///test_usage.db")
+    white_listed_id = db2.get_all_whitelisted_ids()
+    assert user_id in white_listed_id
+
+
+def test_user_message_logger():
+    check = os.path.exists("test_usage.db")
```

Comment:
Please DRY up the repeated setup and teardown of test_usage.db.

See discussion of setup/teardown here: https://stackoverflow.com/questions/6854658/explain-the-setup-and-teardown-python-methods-used-in-test-cases

And here is an example where they used fixtures. https://stackoverflow.com/questions/67255653/how-to-set-up-and-tear-down-a-database-between-tests-in-fastapi